### PR TITLE
Multitenant metrics

### PR DESCRIPTION
**Pull request by briangallagher**
# Issue link
[MGDAPI-2648](https://issues.redhat.com/browse/MGDAPI-2648) 
[MGDAPI-2662](https://issues.redhat.com/browse/MGDAPI-2662) 

# What
Metrics for Multi Tenancy
Scripts for Multi Tenancy

# Verification steps
Build an image based on this branch
`ORG=<yourrepo> make image/build/push`

Install multi tenant RHOAM Operator 
`INSTALLATION_TYPE=multitenant-managed-api make install/sandboxrhoam/operator`
After the Operator Installs change the CSV to reference your new image above 
Then create config including, rhmi cr
`INSTALLATION_TYPE=multitenant-managed-api make install/sandboxrhoam/config`

After RHSSO cluster installs, install the IDP
`NUM_REGULAR_USER=10 NUM_ADMIN=5 REALM_DISPLAY_NAME="DevSandbox" REALM="devsandbox" DEDICATED_ADMIN_PASSWORD=Password1 PASSWORD=Password1 ./scripts/setup-sso-idp.sh`

Go to the terminal of the rhoam operator and run the following to monitor the new metrics:
`while sleep 1; do curl -s localhost:8383/metrics | grep no_; done`

This will monitor for 2 out of the three new metrics `no_tenant_realm` and `no_activated_3scale_tenant_account`
As tenants are created you should see these metrics being populated and subsequently being depleted. 
Login to OpenShift as test-user01, test-user02, test-user03, test-user04, test-user05 to create tenants. 
Verify that the above metrics get populated with the usernames temporarily before being remove completely.

Also verify:
`curl -s localhost:8383/metrics | grep num_tenants`   == 5

### CODE REVIEW

File: pkg/products/threescale/reconciler.go [hunk 1]
```diff
@@ -675,7 +675,7 @@ func (r *Reconciler) reconcileComponents(ctx context.Context, serverClient k8scl
 			"threescale_component_element": "zync-que",
 		})
 
-		if r.installation.Spec.Type == string(integreatlyv1alpha1.InstallationTypeManaged) {
+		if integreatlyv1alpha1.IsRHOAM(integreatlyv1alpha1.InstallationType(r.installation.Spec.Type)) {
```
Comment (KevFan) [inline pkg/products/threescale/reconciler.go:678]: So we only want this set for RHOAM installs and not RHMI ? :thinking:

**Review by KevFan — COMMENTED** [review]
Code looks good, just have a few questions

Verified:
* `no_tenant_realm` and `no_activated_3scale_tenant_account` metrics
[no_acccount.log](https://github.com/integr8ly/integreatly-operator/files/7195058/no_acccount.log)
[no_realm.log](https://github.com/integr8ly/integreatly-operator/files/7195059/no_realm.log)
* Verified `num_tenants`
Screenshot from 2021-09-20 11-19-22
 
Verification notes:
* Needed to manually change csv environment variable, `IDENTITY_PROVIDER_NAME` from `DevSandbox` to `devsandbox` to match the created IDP realm name as otherwise the operator would not correctly find the users to create tenants
Screenshot from 2021-09-20 11-09-55
s

File: Makefile [hunk 2]
```diff
@@ -113,6 +114,16 @@ setup/service_account: kustomize
 setup/git/hooks:
 	git config core.hooksPath .githooks
 
+.PHONY: install/sandboxrhoam/operator
+install/sandboxrhoam/operator:
+	@-oc process -p RHOAM_NAMESPACE=$(SANDBOX_NAMESPACE) -f config/developer-sandbox/sandbox-operator-template.yml | oc create -f - -n $(SANDBOX_NAMESPACE)
```
[the remark this answers is not in the document]
Comment (briangallagher) [reply Makefile:119]: Yeah, I'll add that 👍

File: config/developer-sandbox/sandbox-config-template.yml [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+      operatorsInProductNamespace: false
+      smtpSecret: sandbox-rhoam-smtp
+      deadMansSnitchSecret: sandbox-rhoam-deadmanssnitch
+      pagerDutySecret: sandbox-rhoam-pagerduty
+      priorityClassName: sandbox-pod-priority
+parameters:
+  - name: RHOAM_OPERATOR_NAMESPACE
+    value: "sandbox-rhoam-operator"
+  - name: DMS_URL
+    value: ""
```
[the remark this answers is not in the document]
Comment (briangallagher) [reply config/developer-sandbox/sandbox-config-template.yml:69]: I'll add in a default

File: pkg/products/threescale/reconciler.go (hunk 1 above)
Comment (briangallagher) [reply pkg/products/threescale/reconciler.go:678]: IT should actually be just RHMI

**Comment by KevFan** [comment]
Verified with new changes and code looks good to me :+1: 
```
# HELP no_activated_3scale_tenant_account Users/Tenants who do not have an activated 3Scale account
no_activated_3scale_tenant_account{username="test-user01"} 1
no_activated_3scale_tenant_account{username="test-user02"} 1
no_activated_3scale_tenant_account{username="test-user03"} 1
no_activated_3scale_tenant_account{username="test-user04"} 1
no_activated_3scale_tenant_account{username="test-user05"} 1
# HELP no_tenant_realm Users/Tenants whose realm has not yet provisioned
# TYPE no_tenant_realm gauge
no_tenant_realm{username="test-user01"} 1
no_tenant_realm{username="test-user03"} 1
no_tenant_realm{username="test-user04"} 1
no_tenant_realm{username="test-user05"} 1

sh: v: command not found
sh-4.4$ curl -s localhost:8383/metrics | grep num_tenants
# HELP num_tenants Total number of tenants on the cluster
# TYPE num_tenants gauge
num_tenants{num_tenants="5"} 1
```

/lgtm

**Pull request merged**